slow_conv2d grad_weight: call gemm directly

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65759
* #65758
* #65757
* #65737
* __->__ #65726

This PR isn't strictly necessary since grad_weight doesn't use
parallel_for. However, this does reduce the function overhead and will
make it easier to parallelize in the future.

Differential Revision: [D31257877](https://our.internmc.facebook.com/intern/diff/D31257877)

peterbell10 committed:
slow_conv2d grad_weight: call gemm directly

This PR isn't strictly necessary since grad_weight doesn't use
parallel_for. However, this does reduce the function overhead and will
make it easier to parallelize in the future.

[ghstack-poisoned]

peterbell10 committed:
Update on "slow_conv2d grad_weight: call gemm directly"


This PR isn't strictly necessary since grad_weight doesn't use
parallel_for. However, this does reduce the function overhead and will
make it easier to parallelize in the future.

[ghstack-poisoned]

peterbell10 committed:
Update on "slow_conv2d grad_weight: call gemm directly"


This PR isn't strictly necessary since grad_weight doesn't use
parallel_for. However, this does reduce the function overhead and will
make it easier to parallelize in the future.

[ghstack-poisoned]

peterbell10 committed:
Update on "slow_conv2d grad_weight: call gemm directly"


This PR isn't strictly necessary since grad_weight doesn't use
parallel_for. However, this does reduce the function overhead and will
make it easier to parallelize in the future.

[ghstack-poisoned]

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31257877).

peterbell10 committed:
Update on "slow_conv2d grad_weight: call gemm directly"


This PR isn't strictly necessary since grad_weight doesn't use
parallel_for. However, this does reduce the function overhead and will
make it easier to parallelize in the future.

Differential Revision: [D31257877](https://our.internmc.facebook.com/intern/diff/D31257877)

[ghstack-poisoned]

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31257877).

ngimel commented on aten/src/ATen/native/ConvolutionMM2d.cpp:
```diff
@@ -366,18 +384,24 @@ static void slow_conv2d_backward_weight_out_cpu_template(
 
   auto input = input_.contiguous();
   auto grad_output = grad_output_.contiguous();
+  TORCH_CHECK(grad_weight.is_contiguous(), "slow_conv2d: grad_weight must be contiguous");
```
this is checked earlier, in line 369

ngimel replied:
You also need to check that finput is contiguous?

peterbell10 replied:
Changed to check `finput`.

ngimel commented on aten/src/ATen/native/ConvolutionMM2d.cpp:
```diff
@@ -319,14 +319,32 @@ void slow_conv2d_backward_out_cpu_template(
+  const int64_t n = grad_output.size(0);
+  const int64_t k = grad_output.size(1) * grad_output.size(2);
+
+  constexpr int64_t one = 1;
```
this is unused var (this is a merge-blocking error internally)

peterbell10 replied:
Ah, I added the fixup removing this line to the wrong commit in the stack. Fixed now.

ngimel commented on aten/src/ATen/native/ConvolutionMM2d.cpp:
```diff
@@ -366,18 +384,24 @@ static void slow_conv2d_backward_weight_out_cpu_template(
 
   auto input = input_.contiguous();
```
this looks unnecessary, you only need `input.size(0)` and `input.scalar_type()` below, both can be called on `input_`.

peterbell10 replied:
Agreed, removed.

peterbell10 committed:
Update on "slow_conv2d grad_weight: call gemm directly"


This PR isn't strictly necessary since grad_weight doesn't use
parallel_for. However, this does reduce the function overhead and will
make it easier to parallelize in the future.

Differential Revision: [D31257877](https://our.internmc.facebook.com/intern/diff/D31257877)

[ghstack-poisoned]

ngimel commented:
Can you please rebase this stack? I'm getting the following when I try to import it
```
RuntimeError: We tried to import new changes for D31257876, but this diff has already been landed.  This could occur if you rebased your stack prior to the landed diff being pushed by ShipIt; in this case, wait until this diff is exported with ShipIt, rebase on master, and try again.
```

peterbell10 committed:
Update on "slow_conv2d grad_weight: call gemm directly"


This PR isn't strictly necessary since grad_weight doesn't use
parallel_for. However, this does reduce the function overhead and will
make it easier to parallelize in the future.

Differential Revision: [D31257877](https://our.internmc.facebook.com/intern/diff/D31257877)

[ghstack-poisoned]

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31257877).

peterbell10 committed:
Update on "slow_conv2d grad_weight: call gemm directly"


This PR isn't strictly necessary since grad_weight doesn't use
parallel_for. However, this does reduce the function overhead and will
make it easier to parallelize in the future.

Differential Revision: [D31257877](https://our.internmc.facebook.com/intern/diff/D31257877)

[ghstack-poisoned]

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31257877).
